src/engine/loopingcontrol.cpp

```diff
@@ -835,27 +923,104 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint) {
         } else {
             newloopSamples.end += 2;
         }
-    } else if (newloopSamples.end > samples) {
+    }
+    if (newloopSamples.end > samples) {
         // Do not allow beat loops to go beyond the end of the track
         newloopSamples.end = samples;
+        // Let slotLoopHalve know to adjust beatloop_size
+        m_bBeatloopEndPointIsEndOfTrack = true;
+    } else {
+        m_bBeatloopEndPointIsEndOfTrack = false;
     }
 
-    if (keepStartPoint) {
+    if (keepStartPoint && (enable || m_bLoopingEnabled)) {
```

Comment:
A race condition is not an issue. I have changed the behavior of beatloop_size from how I originally had it at the start of this PR. Now setting beatloop_size to a new value sets a loop of that many beats from the existing loop in point, but does not activate it. This helps see on the waveform where the loop would start and end, allowing the user to set it up before activating it. To activate that pre-setup loop without moving the loop in point, the user can use reloop_toggle (or reloop_cue). Without the check on this line of code, slotBeatLoop would seek to the adjusted loop even though the loop wouldn't be activated.